### PR TITLE
[Webhooks for Asynchronous Events] Authentication Success Event and Failure Event Payload Generation

### PR DESCRIPTION
**Pull request by Sachin-Mamoru**
### Authentication Success
```json
{
  "iss": "https://api.asg.io/t/testasgardeo1/",
  "jti": "bd3294b1-e562-469c-a953-760cb7e98610",
  "iat": 1721148238035,
  "rci": "fd42a0c2-db37-4fec-b07d-b8484b98e1ea",
  "event": {
    "schemas.identity.wso2.org/events/logins/event-type/loginSuccess": {
      "user": {
        "id": "07f47397-2e77-4fce-9fac-41ff509d62de",
        "claims": [
          {
            "uri": "http://wso2.org/claims/username",
            "value": "test1@test.com"
          },
          {
            "uri": "http://wso2.org/claims/mobile",
            "value": "0779438491"
          },
          {
            "uri": "http://wso2.org/claims/emailaddress",
            "value": "sachinmamoru@gmail.com"
          },
          {
            "uri": "http://wso2.org/claims/givenname",
            "value": "Test1"
          },
          {
            "uri": "http://wso2.org/claims/lastname",
            "value": "Test1"
          }
        ],
        "ref": "https://api.asg.io/t/testasgardeo1/scim2/07f47397-2e77-4fce-9fac-41ff509d62de",
        "groups": [
          "testGroup"
        ]
      },
      "tenant": {
        "id": "1",
        "name": "testasgardeo1"
      },
      "userStore": {
        "id": "REVGQVVMVA==",
        "name": "DEFAULT"
      },
      "application": {
        "id": "f27178f9-984b-41df-aee5-372de8ef327f",
        "name": "test-app"
      },
      "authenticationMethods": [
        "BasicAuthenticator",
        "sms-otp-authenticator"
      ]
    }
  }
}
```

### Authentication Failure
```json
{
  "iss": "https://api.asg.io/t/testasgardeo1/",
  "jti": "26b1c1df-3a93-48b7-893b-e8fb82d40c4c",
  "iat": 1721151914333,
  "rci": "a505c8f9-7405-446f-8aac-18c6601feff8",
  "event": {
    "schemas.identity.wso2.org/events/logins/event-type/loginFailed": {
      "user": {
        "id": "07f47397-2e77-4fce-9fac-41ff509d62de",
        "ref": "https://api.asg.io/t/testasgardeo1/scim2/07f47397-2e77-4fce-9fac-41ff509d62de"
      },
      "tenant": {
        "id": "1",
        "name": "testasgardeo1"
      },
      "userStore": {
        "id": "REVGQVVMVA==",
        "name": "DEFAULT"
      },
      "application": {
        "id": "f27178f9-984b-41df-aee5-372de8ef327f",
        "name": "test-app"
      },
      "reason": {
        "id": "SMS-65020",
        "failedStep": {
          "step": 2,
          "idp": "LOCAL",
          "authenticator": "sms-otp-authenticator"
        }
      }
    }
  }
}
```

### Related Issues
- https://github.com/wso2-enterprise/asgardeo-product/issues/25662
- https://github.com/wso2-enterprise/iam-product-management/issues/118

### Related Mails
- The level of customizability & extensibility IAM products should support
- [Architecture] [IS/Asgardeo] Faciliate webhooks to integrate with other systems on asynchornous operations

### CODE REVIEW

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java [hunk 1]
```diff
@@ -17,36 +17,197 @@
...
 
+    private static final Log log = LogFactory.getLog(WSO2LoginEventPayloadBuilder.class);
+    private Organization b2bUserResidentOrganization;
+
     @Override
     public EventPayload buildAuthenticationSuccessEvent(EventData eventData) throws IdentityEventException {
 
-        //TODO: Implement the logic to build the authentication success event payload.
-        return new WSO2AuthenticationSuccessEventPayload();
+        WSO2AuthenticationSuccessEventPayload payload = new WSO2AuthenticationSuccessEventPayload();
```
Comment (malithie) [inline components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: I would recommend using the builder pattern to generate this payload.

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java [hunk 2]
```diff
@@ -17,36 +17,197 @@
...
 import static org.wso2.identity.webhook.common.event.handler.constant.Constants.EVENT_SCHEMA_TYPE_WSO2;
+import static org.wso2.identity.webhook.common.event.handler.util.EventHookHandlerUtils.getReference;
 
 /**
  * WSO2 Login Event Payload Builder.
  */
 public class WSO2LoginEventPayloadBuilder implements LoginEventPayloadBuilder {
 
+    private static final Log log = LogFactory.getLog(WSO2LoginEventPayloadBuilder.class);
+    private Organization b2bUserResidentOrganization;
```
Comment (malithie) [inline components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: what's the requirement to just to keep this as an instance variable.
You should be able to resolve this similar to others.

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java [hunk 3]
```diff
@@ -17,36 +17,197 @@
...
+                        user.addGroup(claimValue);
+                        break;
+                    case Constants.WSO2_CLAIM_ROLES:
+                        user.addRole(claimValue);
+                        break;
+                    case Constants.MULTI_ATTRIBUTE_SEPARATOR:
+                        // Not adding the multi attribute separator to the user claims
+                        break;
+                    case Constants.USER_ORGANIZATION:
+                        setUserOrganization(claimValue);
```
Comment (malithie) [inline components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: I think you can use the user object to set organization and utilize that when building the payload

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java (hunk 1 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Addressed with https://github.com/Sachin-Mamoru/identity-webhook-event-handlers/pull/1/commits/4b46eea549f2da382b2553359f13580e7f3d5212

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java (hunk 3 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Addressed with https://github.com/Sachin-Mamoru/identity-webhook-event-handlers/pull/1/commits/4b46eea549f2da382b2553359f13580e7f3d5212

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java (hunk 2 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Addressed with https://github.com/Sachin-Mamoru/identity-webhook-event-handlers/pull/1/commits/4b46eea549f2da382b2553359f13580e7f3d5212

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java [hunk 4]
```diff
@@ -17,53 +17,207 @@
...
+
+        User user = new User();
+        try {
+            populateUserAttributes(authenticatedUser, user);
+            user.setId(authenticatedUser.getUserId());
+            user.setRef(EventHookHandlerUtils.getInstance().getURL(Constants.SCIM2_ENDPOINT) + "/" +
+                    authenticatedUser.getUserId());
+        } catch (UserIdNotFoundException e) {
+            throw new IdentityEventException("Error while building the event payload", e);
+        }
```
Comment (malithie) [inline components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Better not break the flow based on the fact user id is not found.
Cause user id is introduced later, so we need to understand properly why this error is thrown.
If user id is not present may be we should use a different identifier.

Add a debug log and a todo to denote that we need to verify when user is is thrown and to handle accordingly

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java [hunk 5]
```diff
@@ -17,53 +17,207 @@
...
+                user.setId(authenticatedUser.getUserId());
+                user.setRef(EventHookHandlerUtils.getInstance().getURL(Constants.SCIM2_ENDPOINT) + "/" +
+                        authenticatedUser.getUserId());
+                if (authenticatedUser.getUserStoreDomain() != null) {
+                    userStore = new UserStore(authenticatedUser.getUserStoreDomain());
+                }
+            }
+        } catch (UserIdNotFoundException e) {
+            throw new IdentityEventException("Error while building the event payload", e);
+        }
```
Comment (malithie) [inline components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Same comment above applies

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java [hunk 6]
```diff
@@ -17,53 +17,207 @@
...
+        User user = new User();
+        UserStore userStore = null;
+        try {
+            if (authenticatedUser != null) {
+                user.setId(authenticatedUser.getUserId());
+                user.setRef(EventHookHandlerUtils.getInstance().getURL(Constants.SCIM2_ENDPOINT) + "/" +
+                        authenticatedUser.getUserId());
+                if (authenticatedUser.getUserStoreDomain() != null) {
+                    userStore = new UserStore(authenticatedUser.getUserStoreDomain());
+                }
```
Comment (malithie) [inline components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Take setting user store domain out of this block. It won't be set of user id is not found in this case

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java [hunk 7]
```diff
@@ -17,53 +17,207 @@
...
+        AuthenticationContext authenticationContext = eventData.getAuthenticationContext();
+        AuthenticatedUser authenticatedUser = eventData.getAuthenticatedUser();
+
+        if (authenticatedUser == null) {
+            throw new IdentityEventException("Authenticated user cannot be null.");
+        }
+
+        User user = new User();
+        try {
+            populateUserAttributes(authenticatedUser, user);
```
Comment (malithie) [inline components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Take this out of this block.
If user id is not present attributes will also not present in this case

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java [hunk 8]
```diff
@@ -17,53 +17,207 @@
...
+            String organizationName = WSO2EventHookHandlerDataHolder.getInstance()
+                    .getOrganizationManager().getOrganizationNameById(organizationId);
+            return new Organization(organizationId, organizationName);
+        } catch (OrganizationManagementException e) {
+            if (ERROR_CODE_INVALID_ORGANIZATION_ID.getCode().equals(e.getErrorCode())) {
+                log.debug("Returning an empty string as the organization name as the name is not returned for the given id.");
+            }
+            throw new IdentityEventException(UNEXPECTED_SERVER_ERROR.getCode(),
+                    "Error while retrieving the organization name for the given id.", e);
+        }
```
Comment (malithie) [inline components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Better not throw and break the flow.
Given these are events, it's not significant that the organization name is not present

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java (hunk 4 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Addressed with https://github.com/Sachin-Mamoru/identity-webhook-event-handlers/pull/1/commits/8750eff9ee3a84ff76fd11671bbbc3451fbb0fc8

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java (hunk 6 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Addressed with https://github.com/Sachin-Mamoru/identity-webhook-event-handlers/pull/1/commits/8750eff9ee3a84ff76fd11671bbbc3451fbb0fc8

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java (hunk 8 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Addressed with https://github.com/Sachin-Mamoru/identity-webhook-event-handlers/pull/1/commits/8750eff9ee3a84ff76fd11671bbbc3451fbb0fc8

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java (hunk 7 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Addressed with https://github.com/Sachin-Mamoru/identity-webhook-event-handlers/pull/1/commits/8750eff9ee3a84ff76fd11671bbbc3451fbb0fc8

File: components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java (hunk 5 above)
Comment (Sachin-Mamoru) [reply components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/builder/WSO2LoginEventPayloadBuilder.java]: Addressed with https://github.com/Sachin-Mamoru/identity-webhook-event-handlers/pull/1/commits/8750eff9ee3a84ff76fd11671bbbc3451fbb0fc8

**Pull request closed without merging**